src/write/range.rs

```diff
@@ -0,0 +1,144 @@
+use indexmap::IndexSet;
+use std::ops::{Deref, DerefMut};
+use vec::Vec;
+
+use common::{Format, RangeListsOffset};
+use write::{Address, Error, Result, Section, SectionId, Writer};
+
+define_section!(DebugRanges, RangeListsOffset, "A writable `.debug_ranges` section.");
+define_section!(DebugRngLists, RangeListsOffset, "A writable `.debug_rnglists` section");
+
+define_offsets!(
+    DebugRangesOffsets: RangeId => RangeListsOffset,
+    "The section offsets of all ranges within a `.debug_ranges` section."
+);
+
+define_offsets!(
+    DebugRngListsOffsets: RangeId => RangeListsOffset,
+    "The section offsets of all ranges within a `.debug_rnglists` section."
+);
+
+/// An identifier for a unit in a `UnitTable`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct RangeId(pub usize);
+
+/// A table of ranges that will be stored in a `.debug_ranges` or `.debug_rnglists` section.
+#[derive(Debug, Default)]
+pub struct RangesTable {
+    ranges: IndexSet<RangeList>,
+}
+
+impl RangesTable {
+    /// Add an range to the table and return its code.
+    pub fn add(&mut self, range_list: RangeList) -> RangeId {
+        let (code, _) = self.ranges.insert_full(range_list);
+        // Code must be non-zero
+        RangeId(code)
+    }
+
+    /// Write the range table to the appropriate section for the given DWARF version
+    pub fn write<W: Writer>(
+        &self,
+        w_ranges: &mut DebugRanges<W>,
+        w_rnglists: &mut DebugRngLists<W>,
+        format: Format,
+        version: u16,
+        address_size: u8,
+    ) -> Result<(DebugRangesOffsets, DebugRngListsOffsets)> {
```

Comment:
Maybe we should return a struct:
```rust
struct RangeListOffsets {
   debug_ranges: DebugRangesOffsets,
   debug_rnglists: DebugRngListsOffsets,
}
```
That'll mean one less parameter that needs to be passed around elsewhere.